Kubernetes Executor Documentation Updates (#1811)
astronomer/docs: astro/configure-executors.md

---
sidebar_label: 'Configure an executor'
title: 'Configure a Deployment executor'
id: 'configure-executors'
---

<head>
  <meta name="description" content="Learn how to select and manage Astro executors." />
  <meta name="og:description" content="Learn how to select and manage Astro executors." />
</head>

> The executor determines which worker resources run your scheduled tasks. The difference between executors is primarily based on how tasks are distributed across worker resources.

Reviewer: It's bridged if we just don't mention either tool either tho, right? I almost feel like mentioning both features opens the possibility that the executor somehow works differently in a more nuanced way besides the shared functionality, which is not what we mean to suggest here.

> The executor determines which worker resources run your scheduled tasks. The difference between executors is primarily based on how tasks are distributed across worker resources.

Reviewer: Going with just "The Airflow executor"

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.

2. Click the **Worker Queues** tab and then click **Edit** to edit a worker queue.

3. Configure the following settings:

    - **Max Tasks Per Worker**: The maximum number of tasks that a single worker can run at a time. If the number of queued and running tasks exceeds this number, a new worker is added to run the remaining tasks. This value is equivalent to [worker concurrency](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#worker-concurrency) in Apache Airflow. It is 16 by default.
    - **Worker Count (Min-Max)**: The minimum and maximum number of workers that can run at a time. The number of running workers changes regularly based on **Maximum Tasks per Worker** and the current number of tasks in a queued or running state. By default, the minimum number of workers is 1 and the maximum is 10.
    
4. Click **Update Queue**.

## Manage the Kubernetes executor

> The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) dynamically launches and terminates Pods to run Airflow tasks. The executor starts a new Kubernetes Pod to execute each individual task run, and then shuts down the Pod when the task run completes. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without without task runs

Reviewer 2: Annoying that we're repeating ourselves here but not sure how to avoid. Could we remove some stuff _maybe_?

> The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) dynamically launches and terminates Pods to run Airflow tasks. The executor starts a new Kubernetes Pod to execute each individual task run, and then shuts down the Pod when the task run completes. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without without task runs

Reviewer 2: ```suggestion
The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) dynamically launches and terminates Pods to run Airflow tasks. The executor starts a new Kubernetes Pod to execute each individual task run, and then shuts down the Pod when the task run completes. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without without task runs.
```

> The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) dynamically launches and terminates Pods to run Airflow tasks. The executor starts a new Kubernetes Pod to execute each individual task run, and then shuts down the Pod when the task run completes. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without without task runs.

Reviewer: ```suggestion
The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) dynamically launches and terminates Pods to run Airflow tasks. The executor starts a new Kubernetes Pod to execute each individual task run, and then shuts down the Pod when the task run completes. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without task runs.
```

By default, each task on Astro runs in a dedicated Kubernetes Pod with 1 CPU and 512Mi of memory. These Pods run on a cloud worker node, which can run multiple worker Pods at once. If a worker node can't run any more Pods, Astro automatically provisions a new worker node to begin running any queued tasks in new Pods.

> ### Customize the Kubernetes worker Pod for a task

Reviewer 2: ```suggestion
### Customize a Kubernetes Pod for a task
```

Not sure we call them "worker Pods" but maybe i'm wrong

Reviewer: I updated this to plural as it's likely customers will assign multiple tasks to a single Pod spec

:::warning

> While you can technically customize all values for a worker Pod, Astronomer recommends against configuring complex Kubernetes infrastructure in your Pods such as sidecars. These configurations have not been tested by Astronomer.

Reviewer 2: Another example would be good here in the future

> While you can technically customize all values for a worker Pod, Astronomer recommends against configuring complex Kubernetes infrastructure in your Pods such as sidecars. These configurations have not been tested by Astronomer.

Reviewer 2: ```suggestion
While you can customize all values for a worker Pod, Astronomer does not recommend configuring complex Kubernetes infrastructure in your Pods, such as sidecars. These configurations have not been tested by Astronomer.
```

:::

You can configure different custom worker Pods to override the default Astro worker Pod on a per-task basis. You might complete this setup to change how many resources the Pod uses.

1. Add the following import to your DAG file:

    ```sh
    from kubernetes.client import models as k8s
    ```

2. Add a `pod_override` configuration to the DAG file containing the task. See the [`kubernetes-client`](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1Container.md) GitHub for a list of all possible settings you can include in the configuration.
3. Specify the `pod_override` in the task's parameters.

See [Manage task CPU and memory](#manage-task-cpu-and-memory) for an example `pod_override` configuration. 

> ### Configure limits and requests for worker Pod CPU and memory

Reviewer: ```suggestion
### Example Pod configuration: Configure limits and requests for worker Pod CPU and memory
```

One of the most common use cases for customizing a Kubernetes worker Pod is to request a specific amount of resources for a task. When requesting resources, make sure that your requests don't exceed the available resources in your current [Pod worker node type](#change-the-pod-worker-node-type).

> The following example shows how you can use a `pod_override` configuration in your DAG code to request custom resources for a task:

Reviewer 2: confused about the relationship between these 2 sections -- are they both telling me how to do CPU and memory? Is one a pre-requisiste?

Reviewer: I had "Example" at the start of this topic before, but it was removed during review. I think it's worth adding back
